File: src/main/java/jp/kentan/minecraft/neko_core/component/zone/Zone.java
package jp.kentan.minecraft.neko_core.component.zone;

import org.bukkit.ChatColor;
import org.bukkit.World;

public class Zone {

    public String ID, NAME;
    public final ZoneType TYPE;
    public final World WORLD;
    public final int OWNER_LIMIT;
    public final int RENTAL_DAYS;
    public final double PRICE_RATE, PRICE_RATE_GAIN, SELL_GAIN;

    public String BUY_RENTAL_RULE, SELL_RULE;

    public Zone(String id,
                String name,
                ZoneType type,
                World world,
                int ownerLimit,
                int rentalDays,
                double priceRate,
                double priceRateGain,
                double sellGain,
                String buyRentalRule,
                String sellRule) {
        ID = id;
        NAME = ChatColor.translateAlternateColorCodes('&', name);
        TYPE = type;
        WORLD = world;
        OWNER_LIMIT = ownerLimit;
        RENTAL_DAYS = rentalDays;
        PRICE_RATE = priceRate;
        PRICE_RATE_GAIN = priceRateGain;
        SELL_GAIN = sellGain;

        BUY_RENTAL_RULE = (buyRentalRule != null) ? ChatColor.translateAlternateColorCodes('&', buyRentalRule) : null;
        SELL_RULE       = (sellRule      != null) ? ChatColor.translateAlternateColorCodes('&', sellRule     ) : null;
    }
}

File: src/main/kotlin/jp/kentan/minecraft/nekocore/data/model/Zone.kt
package jp.kentan.minecraft.nekocore.data.model

import org.bukkit.World
import kotlin.math.max
import kotlin.math.min

data class Zone(
    val id: String,
    val name: String,
    val type: Type,
    val world: World,
    val ownedLimit: Int,
    val rentalDays: Int,
    val priceRate: Double,
    val priceRateGain: Double,
    val sellGain: Double,
    val buyRentalRule: String?,
    val sellRule: String?
) {
    enum class Type(
        val displayName: String
    ) {
        BUY_UP("買い切り"),
        RENTAL("レンタル");
    }

    fun calcAreaPurchasePrice(size: Int, ownedAreaCount: Int, isOwned: Boolean = false): Double {
        val ownedCount = if (isOwned) max(0, ownedAreaCount - 1) else ownedAreaCount
        val count = min(ownedCount, ownedLimit - 1)
        val rate = if (count > 0) priceRate * max(1.0, priceRateGain * count) else priceRate
        return rate * size
    }

    fun calcAreaSellPrice(purchasedPrice: Double): Double = sellGain * purchasedPrice
}